modules/basics/src/main/java/com/opengamma/basics/currency/FxMatrix.java

```diff
@@ -0,0 +1,557 @@
+   * using the rates in this matrix.
+   *
+   * @param amount  the currency amount, not null
+   * @param ccy  the currency to convert all entries to
+   * @return the amount converted to the requested currency
+   */
+  public CurrencyAmount convert(CurrencyAmount amount, Currency ccy) {
+    ArgChecker.notNull(amount, "amount");
+    ArgChecker.notNull(ccy, "ccy");
+    Currency originalCcy = amount.getCurrency();
+    // Only do conversion if we need to
+    return originalCcy == ccy ? amount : CurrencyAmount.of(ccy, amount.getAmount() * getRate(originalCcy, ccy));
+  }
+
+  /**
+   * Convert a multiple currency amount into a amount in the specified currency
+   * using the rates in this matrix.
+   *
+   * @param amount  the multiple currency amount, not null
+   * @param ccy  the currency to convert all entries to
+   * @return the total amount in the requested currency
+   */
+  public CurrencyAmount convert(MultiCurrencyAmount amount, Currency ccy) {
+    ArgChecker.notNull(amount, "amount");
+    ArgChecker.notNull(ccy, "ccy");
+
+    // We could do this using the currency amounts but to
+    // avoid creating extra objects we'll use doubles
+    double total = amount.getAmounts()
+        .stream()
+        .mapToDouble(ca -> ca.getAmount() * getRate(ca.getCurrency(), ccy))
+        .sum();
+    return CurrencyAmount.of(ccy, total);
+  }
+
+  /**
+   * Merge the entries from the other matrix into this one. The other matrix
+   * should have at least one currency in common with this one.
+   * The additional currencies from the other matrix are added one by one
+   * and the exchange rate data created is coherent with some data in this
+   * matrix. If the data in the two matrices are not coherent then there is
+   * no guarantee which data will be used and the final result may be incoherent.
+   *
+   * @param other  the matrix to be merged into this one
+   * @return a new matrix containing the rates from this matrix
+   *   plus any rates for additional currencies from the other matrix
+   */
+  public FxMatrix merge(FxMatrix other) {
+    return toBuilder().merge(other.toBuilder()).build();
+  }
+
+  /**
+   * Returns an immutable set containing the currencies held within this matrix.
+   *
+   * @return the currencies in this matrix
+   */
+  public ImmutableSet<Currency> getCurrencies() {
+    return currencies.keySet();
+  }
+
+  @Override
+  public String toString() {
+    return getCurrencies() + " - " + Stream.of(rates).map(Arrays::toString).collect(Collectors.joining());
+  }
+
+  @Override
+  public int hashCode() {
+    return 31 * currencies.hashCode() + Arrays.deepHashCode(rates);
+  }
+
+  @Override
+  public boolean equals(final Object obj) {
+    if (this == obj) {
+      return true;
+    }
+    if (obj == null || getClass() != obj.getClass()) {
+      return false;
+    }
+    FxMatrix other = (FxMatrix) obj;
+    return currencies.equals(other.currencies) && Arrays.deepEquals(rates, other.rates);
+  }
+
+  /**
```

Comment:
It looks like the Javadoc for this method and the one below have been switched round.